File: graph_algorithms/week3_shortest_paths/1_flight_segments/bfs.py
from queue import Queue
from typing import List


def distance(adj: List[List[int]], s: int, t: int) -> int:  # noqa: C901
    dist = [-1] * len(adj)
    dist[s] = 0
    queue: Queue = Queue()
    queue.put(s)
    while dist[t] == -1 and not queue.empty():
        u = queue.get()
        for v in adj[u]:
            if dist[v] == -1:
                dist[v] = dist[u] + 1
                queue.put(v)
    return dist[t]


if __name__ == "__main__":
    n, m = list(map(int, input().split()))

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)

    u, v = map(int, input().split())
    print(distance(adj, u - 1, v - 1))

File: graph_algorithms/week3_paths_in_graphs1/1_flight_segments/bfs.cpp
#include <iostream>
#include <queue>
#include <vector>

using std::queue;
using std::vector;

int distance(vector<vector<int> > &adj, int s, int t) {
  vector<int> dist(adj.size(), -1);
  dist[s] = 0;
  queue<int> q;
  q.push(s);
  while (dist[t] == -1 && !q.empty()) {
    int u = q.front();
    q.pop();
    for (int v : adj[u])
      if (dist[v] == -1) {
        dist[v] = dist[u] + 1;
        q.push(v);
      }
  }
  return dist[t];
}

int main() {
  int n, m;
  std::cin >> n >> m;

  vector<vector<int> > adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }

  int u, v;
  std::cin >> u >> v;
  std::cout << distance(adj, u - 1, v - 1) << std::endl;
}
